### PR TITLE
Relax pattern for softmax matching

### PR DESCRIPTION
**Pull request by ThomasRaoux**

### CODE REVIEW

File: llvm-external-projects/iree-dialects/lib/Transforms/TransformMatchers.cpp [hunk 1]
```diff
@@ -499,34 +501,33 @@ transform_ext::StructuredOpMatcher &transform_ext::StructuredOpMatcher::input(
...
-      return false;
-    Operation *producer = linalgOp.getDpsInputOperand(updatedProducerPosition)
-                              ->get()
-                              .getDefiningOp();
-    if (!producer)
-      return false;
+    // LLVM_DEBUG(DBGS() << "input operand " << position
+    //                   << " is the same value as operand "
+    //                   << producerPosition.value << " of the input "
+    //                   << producerPosition.value << " producer");
```
Comment (ftynse) [inline llvm-external-projects/iree-dialects/lib/Transforms/TransformMatchers.cpp]: Can we keep at least something?

File: llvm-external-projects/iree-dialects/lib/Transforms/TransformMatchers.cpp [hunk 2]
```diff
@@ -499,34 +501,33 @@ transform_ext::StructuredOpMatcher &transform_ext::StructuredOpMatcher::input(
...
+    linalg::LinalgOp producerOp = linalgOp;
+    for (size_t i = 0, e = producerPosition.value.size() - 1; i < e; i++) {
```
Comment (ftynse) [inline llvm-external-projects/iree-dialects/lib/Transforms/TransformMatchers.cpp]: Nit: something like `int64_t pos : llvm::drop_front(producerPosition.value)` would be more idiomatic here.

File: llvm-external-projects/iree-dialects/include/iree-dialects/Transforms/TransformMatchers.h [hunk 3]
```diff
@@ -136,13 +136,13 @@ struct IsProjected : public SingleValuePredicateParam<int64_t> {
...
 /// Predicate tag indicating that the operand is a special float constant.
-struct ConstantFloatMin {};
+struct ConstantFloatMinOrMinusInf {};
 struct ConstantFloatZero {};
 
 /// Predicate indicating that the operand is the same value as its producer's
 /// operand.
 struct SameOperandAsProducer
-    : public SingleValuePredicateParam<std::pair<int64_t, int64_t>> {
+    : public SingleValuePredicateParam<SmallVector<int64_t>> {
```
Comment (ftynse) [inline llvm-external-projects/iree-dialects/include/iree-dialects/Transforms/TransformMatchers.h]: Could you document the semantics of values in this vector?

File: llvm-external-projects/iree-dialects/lib/Transforms/TransformMatchers.cpp [hunk 4]
```diff
@@ -758,11 +759,21 @@ transform_ext::StructuredOpMatcher::singleOpWithCanonicaleArgs(
...
-    for (auto [index, operand] : llvm::enumerate(innerOp->getOperands())) {
-      auto arg = dyn_cast<BlockArgument>(operand);
-      if (!arg || arg.getParentBlock() != linalgOp.getBlock() ||
-          arg.getArgNumber() != index)
+    if (commutative && innerOp->getNumOperands() == 2) {
+      auto arg0 = dyn_cast<BlockArgument>(innerOp->getOperand(0));
+      auto arg1 = dyn_cast<BlockArgument>(innerOp->getOperand(1));
+      if (!arg0 || !arg1)
+        return false;
+      if (!((arg0.getArgNumber() == 0 && arg1.getArgNumber() == 1) ||
```
Comment (ftynse) [inline llvm-external-projects/iree-dialects/lib/Transforms/TransformMatchers.cpp]: This no longer checks whether `arg0` and `arg1` are arguments of the `linalgOp` block, and will catch, e.g., function arguments.

File: llvm-external-projects/iree-dialects/include/iree-dialects/Transforms/TransformMatchers.h [hunk 5]
```diff
@@ -505,13 +505,18 @@ class StructuredOpMatcher : public CapturingOpMatcher {
...
+    return singleOpWithCanonicaleArgs(OpType::getOperationName(), commutative);
   }
-  StructuredOpMatcher &singleOpWithCanonicaleArgs(StringRef opname);
+  StructuredOpMatcher &singleOpWithCanonicaleArgs(StringRef opname,
+                                                  bool commutative);
   /// Check if the op is a linalg of with a single float reciprocal op.
   StructuredOpMatcher &isFloatReciprocal();
+  /// Check if the op is a linalg of with a region containing only a yield op
+  /// using block arguments in order.
+  StructuredOpMatcher &passThroughtOp();
```
Comment (ftynse) [inline llvm-external-projects/iree-dialects/include/iree-dialects/Transforms/TransformMatchers.h]: Nit: "through" doesn't have a t in the end.

File: llvm-external-projects/iree-dialects/include/iree-dialects/Transforms/TransformMatchers.h (hunk 3 above)
Comment (ftynse) [reply llvm-external-projects/iree-dialects/include/iree-dialects/Transforms/TransformMatchers.h]: Note that #12168 removes this completely in favor of value matchers.

File: llvm-external-projects/iree-dialects/lib/Transforms/TransformMatchers.cpp (hunk 4 above)
Comment (ThomasRaoux) [reply llvm-external-projects/iree-dialects/lib/Transforms/TransformMatchers.cpp]: thanks, added back

File: llvm-external-projects/iree-dialects/include/iree-dialects/Transforms/TransformMatchers.h (hunk 3 above)
Comment (ThomasRaoux) [reply llvm-external-projects/iree-dialects/include/iree-dialects/Transforms/TransformMatchers.h]: awesome, thanks Alex

**Review by ftynse — APPROVED** [review]

**Pull request merged**